### PR TITLE
[c10] add an option to pg_config split share

### PR DESCRIPTION
**Pull request by shuqiangzhang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130877

Summary:
context is: #129865
We want to give users an option to not share comms resouces so that
comm opts can overlap
Test Plan:
Augmentd UT

Tags:

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

**Comment by shuqiangzhang** [comment]
cc @nvcastet

File: torch/csrc/distributed/c10d/init.cpp [hunk 1]
```diff
@@ -2784,6 +2785,7 @@ Example::
     >>> nccl_options.config.cga_cluster_size = 2
     >>> nccl_options.config.max_ctas = 4
     >>> nccl_options.config.min_ctas = 2
+    >>> nccl_options.config.split_share = 1
```
Comment (nvcastet) [inline torch/csrc/distributed/c10d/init.cpp:2788]: I would put `split_share = 0` for the example here since you plan to put the default to `1`.

**Comment by shuqiangzhang** [comment]
> > cc @nvcastet, cannot init config.splitShare = 1 after `ncclConfig_t config = NCCL_CONFIG_INITIALIZER;' due to compile time error: ‘config’ does not name a type
> 
> Yes you need to put it inside the `#ifdef` since `config` is only declared there. Inside that `#ifdef` you would also need to add a nested `#ifdef NCCL_HAS_COMM_SPLIT` to make sure `splitShare` is present in that NCCL version.

I did exactly both yesterday, the compiler was still not happy

File: torch/csrc/distributed/c10d/init.cpp (hunk 1 above)
Comment (shuqiangzhang) [reply torch/csrc/distributed/c10d/init.cpp:2788]: In this version of code, split_share = 0 is the default

File: test/distributed/test_c10d_nccl.py [hunk 2]
```diff
@@ -2796,6 +2796,7 @@ def test_pass_nccl_options_config(self):
         pg_opts.config.min_ctas = 2
         pg_opts.config.cga_cluster_size = 2
         pg_opts.config.net_name = "Socket"
+        pg_opts.config.split_share = 1
```
Comment (fduwjj) [inline test/distributed/test_c10d_nccl.py:2799]: Not necessary to do it in this PR. Do you think we want to add some docs in distributed_c10d.py about NCCL config in the option? Currently the only info we have is by reading through cpp code and unit test, IIUC.

Comment (shuqiangzhang) [reply test/distributed/test_c10d_nccl.py:2799]: more clearer doc is always welcome ^_^

**Comment by shuqiangzhang** [comment]
cc @mengph , if this PR is accepted as is (splitShare=0) would be the default consistent with NCCL, so you will have to set the pg_opts.config.split_share = 1 and pass it to init_process_group if you want the 'share' behavior

File: test/distributed/test_c10d_nccl.py (hunk 2 above)
Comment (wconstab) [reply test/distributed/test_c10d_nccl.py:2799]: but at least we should put a comment saying what this does?  i'm actually not clear on when i would change it-

is it this? or something else?
1= share resources so less memory and sockets are used (unclear how much savings here vs the tradeoff of 0)
0= can overlap p2p comms but waste more communicators and init time?

Comment (shuqiangzhang) [reply test/distributed/test_c10d_nccl.py:2799]: All those are NCCL specific configs, which are documented by NCCL as the source of truth and referred in the Pybind file (init.cpp) line 2739: https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/types.html#ncclconfig-t. 
image. Specifically, for this config, it explains as "Specify whether to share resources with child communicator during communicator split. Set the value of splitShare to 0 or 1. The default value is 0. When the parent communicator is created with splitShare=1 during ncclCommInitRankConfig, the child communicator can share internal resources of the parent during communicator split. Split communicators are in the same family. When resources are shared, aborting any communicator can result in other communicators in the same family becoming unusable. Irrespective of whether sharing resources or not, users should always abort/destroy all no longer needed communicators to free up resources.

"

**Review by fduwjj — APPROVED** [review]
Stamped now and as we have discussed offline, it will be great if we can surface the doc for the nccl config a little bit more.

**Pull request closed without merging**